contracts/utils/cryptography/signers/SignerWebAuthn.sol

```diff
@@ -0,0 +1,60 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.24;
+
+import {SignerP256} from "./SignerP256.sol";
+import {WebAuthn} from "./WebAuthn.sol";
+
+/**
+ * @dev Implementation of {SignerP256} that supports WebAuthn authentication assertions.
+ *
+ * This contract enables signature validation using WebAuthn authentication assertions,
+ * leveraging the P256 public key stored in the contract. It allows for both WebAuthn
+ * and raw P256 signature validation, providing compatibility with both signature types.
+ *
+ * The signature is expected to be an abi-encoded {WebAuthn-WebAuthnAuth} struct.
+ *
+ * Example usage:
+ *
+ * ```solidity
+ * contract MyAccountWebAuthn is Account, SignerWebAuthn, Initializable {
+ *     function initialize(bytes32 qx, bytes32 qy) public initializer {
+ *         _setSigner(qx, qy);
+ *     }
+ * }
+ * ```
+ *
+ * IMPORTANT: Failing to call {_setSigner} either during construction (if used standalone)
+ * or during initialization (if used as a clone) may leave the signer either front-runnable or unusable.
+ */
+abstract contract SignerWebAuthn is SignerP256 {
+    /**
+     * @dev Validates a raw signature using the WebAuthn authentication assertion.
+     *
+     * In case the signature can't be validated, it falls back to the
+     * {SignerP256-_rawSignatureValidation} method for raw P256 signature validation by passing
+     * the raw `r` and `s` values from the signature.
+     */
+    function _rawSignatureValidation(
+        bytes32 hash,
+        bytes calldata signature
+    ) internal view virtual override returns (bool) {
+        (bytes32 qx, bytes32 qy) = signer();
+
+        return
+            WebAuthn.verifyMinimal(abi.encodePacked(hash), _toWebAuthnSignature(signature), qx, qy) ||
+            super._rawSignatureValidation(hash, signature);
+    }
+
+    /// @dev Non-reverting version of signature decoding.
+    function _toWebAuthnSignature(bytes calldata signature) private pure returns (WebAuthn.WebAuthnAuth memory auth) {
+        bool decodable;
+        assembly ("memory-safe") {
+            let offset := calldataload(signature.offset)
+            // Validate the offset is within bounds and makes sense for a WebAuthnAuth struct
+            // A valid offset should be 32 and point to data within the signature bounds
+            decodable := and(eq(offset, 32), lt(add(offset, 0x80), signature.length))
```

Comment:
Assuming standard abi encoding, we know the struct will occupy at least 8 slots (length of 0x100) since there will be an offset and at least a length slot for both the bytes and string params. Should we be stricter?

Reply:
In reality this check is to avoid [ERC7739 detection](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/f45e9158b7d2c46a89f6f974c943dc51fbdd35ac/contracts/utils/cryptography/signers/ERC7739.sol#L43). It's been a bit of an issue across signers to return when the `signature` is empty, so I don't have a strong opinion but I think it's best to be relaxed if there's no security concern. People can check on top if they want to be more strict